docs/plugins/pluggable_authentication.md

```diff
@@ -83,3 +83,165 @@ Since PAM requires the user's password in plaintext, iRODS relies on TLS encrypt
+
+#### How to deal with errors
+
+If an error occurs such that the correct inputs cannot be provided from the client side to the server side operation, the authentication should be considered failed and the flow should be aborted. If the `authenticate` API returns a non-zero return value, the authentication should be considered a failure and the flow should be aborted.
+
+### Implementing authentication plugins
+
+!!! Note
+    This section will likely only be of interest to client library developers.
+
+#### Native authentication
+
+The native authentication plugin has two server-side operations which must be called in order. The inputs and outputs from each are described below.
+
+**Step 1: `auth_agent_auth_request`**
+
+This operation - like all server-side plugin operations - requires a "scheme" in the request payload so that the server knows which authentication plugin to load. Otherwise, this operation has no requirements in terms of keys to include in the request payload. The request payload should minimally contain the following:
+```json
+{
+    "scheme": "native"
+}
+```
+
+The response payload from the server will minimally contain the following:
+```json
+{
+    "scheme": "native",
+    "request_result": "64_random_bytes_encoded_as_a_string____________________________"
+}
+```
+The `request_result` is a randomly-generated byte-string that the server also preserves and uses to verify the client with which it is communicating.
+
+**Step 2: `auth_agent_auth_response`**
+
+This operation requires 3 keys: `digest`, `user_name`, and `zone_name`.
+
+`user_name` and `zone_name` are the username and zone name for the user being authenticated. A user of the specified name belonging the specified zone must exist in the local zone in order for authentication to succeed.
+
+`digest` is a base64-encoded MD5 hash of the user's password. The client should construct the MD5 hash by concatenating two strings:
+
+1. The randomly-generated string from the server stored in `request_result`
+2. The user's plaintext password
+
+How the plugin gets the user's password is entirely up to the library developer. For reference, the C++ client-side authentication plugin has a few ways of obtaining the password:
+
+- The `a_pw` key in the request payload
+- Prompting the user and receiving the passwords from `stdin`
+- Reading from an irodsA file
+
+For example, if the user's password is "apass", the string to hash looks like this:
+```
+64_random_bytes_encoded_as_a_string____________________________apass
+```
+The concatenated string should be used as input to a MD5 hashing function, and the resulting byte-string should be base64-encoded. The resulting string should be stored in the `digest` key.
+
+The request payload should minimally contain the following:
+```json
+{
+    "scheme": "native",
+    "user_name": "<user_name>",
+    "zone_name": "<zone_name>",
+    "digest": "<base64-encoded string of MD5 hash>"
```

Comment:
Consider replacing the hyphens and spaces with underscores in the value for `digest`. Or remove the underscores from the user name and zone name values.

Reply:
Removed underscores